java/code/src/com/redhat/rhn/frontend/xmlrpc/packages/PackagesHandler.java

```diff
@@ -597,4 +605,110 @@ public String getPackageUrl(User loggedInUser, Integer pid) {
         return toReturn;
     }
 
+    /**
+     * Add a package lock.
+     * @param loggedInUser The current user
+     * @param sid The System ID
+     * @param pid The Package ID
+     * @return 1 on success.
+
+     * @xmlrpc.doc Add a package lock
+     * @xmlrpc.param #session_key()
+     * @xmlrpc.param #param("int", "system_id")
+     * @xmlrpc.param #param("int", "package_id")
+     * @xmlrpc.returntype #return_int_success()
+     *
+     */
+    public int addLock(User loggedInUser, Integer sid, Integer pid) {
+        Package pkg = lookupPackage(loggedInUser, pid);
+        DataResult<PackageListItem> lockedPackagesResult =
+                PackageManager.systemLockedPackages(sid.longValue(), null);
+
+        Set<Package> pkgsAlreadyLocked = lockedPackagesResult.stream().map(
+                x -> PackageManager.lookupByIdAndUser(x.getPackageId(), loggedInUser)).collect(Collectors.toSet());
+
+        if (pkgsAlreadyLocked.contains(pkg)) {
+            logger.warn("Package " + pkg.getNameEvra() + " with pid " + pkg.getId() + "already locked");
+            return 0;
+        }
+
+        PackageManager.setPendingStatusOnLockedPackages(List.of(pkg), PackageManager.PKG_PENDING_LOCK);
+        PackageManager.lockPackages(sid.longValue(), Set.of(pkg));
```

Comment:
as discussed, it needs to be scheduled. And schedule functions all belong to the SystemHander.
I would suggest to move all these functions to the system handler and maybe reduce it to 1 schedule function only for changing the locks.